### PR TITLE
TEST: improve coverage for clustering utils helpers and edge cases

### PR DESCRIPTION
**Pull request by Piyush-t24**
## Overview

Related to #3690

This PR improves test coverage for `shap/utils/_clustering.py` by expanding `tests/utils/test_clustering.py`.

Description of the changes proposed in this pull request:

- Added focused tests for numba-accelerated helpers:
  - `_pt_shuffle_rec`
  - `delta_minimization_order`
  - `_reverse_window`
  - `_reverse_window_score_gain`
  - `_mask_delta_score`
- Added/expanded tests for public clustering APIs:
  - `partition_tree`
  - `partition_tree_shuffle`
  - `hclust`
  - `hclust_ordering`
  - `xgboost_distances_r2`
- Covered deterministic happy paths, edge cases, and branch-specific behavior.
- Kept PR scope limited to a single test file, as requested in #3690.

Commands run locally:

- `python -m pytest tests/utils/test_clustering.py -q`
- `python -m coverage run -m pytest tests/utils/test_clustering.py -q`
- `python -m coverage report -m shap/utils/_clustering.py`
- `python -m pre_commit run --files tests/utils/test_clustering.py`

Current local module coverage for `shap/utils/_clustering.py`: **100%**.

AI usage disclosure:

I used AI assistance (Cursor) to help brainstorm test scenarios, draft/refine test cases, and improve branch coverage. I manually reviewed all changes, ran the checks locally, and take responsibility for the final code and PR content.

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

**Comment by Piyush-t24** [comment]
@CloseChoice Hi Tobias, I’ve opened this scoped PR for #3690 (single-file test coverage for shap/utils/_clustering.py). Could you please take a look when you get a chance? Thanks!

File: tests/utils/test_clustering.py [hunk 1]
```diff
@@ -1,42 +1,201 @@
...
-    # just check if clustered ran successfully (using xgboost_distances_r2)
-    clustered = hclust(X, y, linkage=linkage, random_state=0)
-    assert isinstance(clustered, np.ndarray)
-    assert clustered.shape == (1, 4)
 
-    # Check clustering runs if y=None (using scipy metrics)
-    clustered = hclust(X, linkage=linkage, random_state=0)
-    assert isinstance(clustered, np.ndarray)
-    assert clustered.shape == (1, 4)
+class TestNumbaHelpers:
```
Comment (CloseChoice) [inline tests/utils/test_clustering.py]: Pls let's not use test classes, just functions

File: tests/utils/test_clustering.py [hunk 2]
```diff
@@ -1,42 +1,201 @@
...
+
+        # Public wrapper path.
+        np.random.seed(42)
+        wrapper_mask = np.array([True, True, False], dtype=bool)
+        out = np.empty(wrapper_mask.sum(), dtype=np.int64)
+        partition_tree_shuffle(out, wrapper_mask, tree)
+        assert set(out.tolist()) == {0, 1}
+
+
+class TestPublicClusteringAPI:
```
Comment (CloseChoice) [inline tests/utils/test_clustering.py]: same

**Review by CloseChoice — CHANGES_REQUESTED** [review]
thanks for the PR. Are all tests needed? Could we trim down some of them? Pls also refactor to just use function style tests

**Comment by Piyush-t24** [comment]
@CloseChoice  Thanks for the feedback! I've addressed both points:

Refactored all test classes to plain function-style tests
Trimmed redundant cases — down to 10 focused tests

Coverage is now at 100% on shap/utils/_clustering.py . All pre-commit checks pass. Please take another look! 🙏
